src/Raven.Server/Rachis/Leader.cs

```diff
@@ -777,42 +771,43 @@ private void EmptyQueue()
                                     }
                                     else
                                     {
-                                        if (result != null)
+                                        if (result != null && cmd.CtxResultWriter != null)
                                         {
-                                            result = GetConvertResult(cmd.Command)?.Apply(result) ?? cmd.Command.FromRemote(result);
+                                            cmd.CtxResultWriter.CopyResult(result);
+                                            result = null;
                                         }
 
                                         cmd.Tcs.TrySetResult(Task.FromResult<(long, object)>((index, result)));
                                     }
-                                    list.Remove(cmd.Tcs);
                                     continue;
                                 }
                             }
                             else
                             {
+                                var djv = cmd.Command.ToJson(context);
+                                var cmdJson = context.ReadObject(djv, "raft/command");
                                 index = _engine.InsertToLeaderLog(context, Term, cmdJson, RachisEntryFlags.StateMachineCommand);
                             }
+                            list.Add(cmd.Tcs);
 
-                            if (_entries.TryGetValue(index, out var state))
-                            {
-                                tasks.Add(state.TaskCompletionSource.Task);
-                            }
-                            else
+                            if (_entries.TryGetValue(index, out var state) == false)
                             {
                                 var tcs = new TaskCompletionSource<(long, object)>(TaskCreationOptions.RunContinuationsAsynchronously);
-                                tasks.Add(tcs.Task);
-                                state = new
-                                    CommandState
+                                state = new CommandState
                                 // we need to add entry inside write tx lock to avoid
                                 // a situation when command will be applied (and state set)
                                 // before it is added to the entries list
                                 {
                                     CommandIndex = index,
                                     TaskCompletionSource = tcs,
-                                    ConvertResult = GetConvertResult(cmd.Command),
                                 };
                                 _entries[index] = state;
                             }
+
+                            if (cmd.CtxResultWriter != null)
+                                state.WriteResultAction += cmd.CtxResultWriter.CopyResult;
```

Comment:
please add a comment why we need it... especially that we can have more than one delegate here

Reply:
done